File: dataStructures/linkedList.py
class Node:

    def __init__(self, data):
        self.data = data
        self.nextNode = None

    def getData(self):
        return self.data

    def getNext(self):
        return self.nextNode

    def setData(self, data):
        self.data = data

    def setNext(self, node):
        self.nextNode = node

class DoublyLinkedNode(Node):

    def __init__(self, data):
        self.data = data
        self.nextNode = None
        self.previousNode = None

    def getPrevious(self):
        return self.previousNode

    def setPrevious(self, node):
        self.previousNode = node


class DoublyLinkedList:

    def __init__(self):
        self.head = None

    def add(self, item):
        temp = DoublyLinkedNode(item)
        if (self.head != None):
#            print "here"
            self.head.setPrevious(temp)
        temp.setNext(self.head)
        self.head = temp

    def remove(self, item):
        found = False
        current = self.head
        
        while (current != None and not found):
            if (current.getData() == item):
                found = True
            else:
                current = current.getNext()

        if (current.previous == None):
            # deleting first item from list
            self.head = current.getNext()
            if (self.head != None):
                self.head.previous = None
        elif (current == None):
            # item not found, return false
            return False
        else:
            previous = current.getPrevious()
            following = current.getNext()
            previous.setNext(following)
            if (following != None):
                following.setPrevious(previous)

    def search(self, item):
        current = self.head
        found = False

        while (current != None and not found):
            if (current.getData() == item):
                found = True
            else:
                current = current.getNext()

        return found

    def isEmpty(self):
        return self.head == None

    def length(self):
        current = self.head
        count = 0

        while (current != None):
            count += 1
            current = current.getNext()

        return count

    def append(self, item):
        current = self.head
        previous = None

        while (current != None):
            previous = current
            current = current.getNext()

        if (previous == None):
            # appending to empty list
            self.add(item)
        else:
            temp = DoublyLinkedNode(item)
            previous.setNext(temp)
            temp.setPrevious(previous)

    def index(self, item):
        current = self.head
        pos = 0
        found = False

        while (current != None and not found):
            if (current.getData() == item):
                found = True
            else:
                pos += 1
                current= current.getNext()

        if (found):
            return pos
        else:
            return False

    def insert(self, pos, item):
        current = self.head
        currentPos = 0

        while (current != None and currentPos != pos):
            current = current.getNext()
            currentPos += 1

        if (current == None):
            return False

        previous = current.getPrevious()
        temp = DoublyLinkedNode(item)
        previous.setNext(temp)
        temp.setNext(current)
        temp.setPrevious(previous)
        current.setPrevious(temp)

    def pop(self, pos=None):
        if (pos == None):
            pos = self.length()-1

        current = self.head
        currentPos = 0

        while (current != None and currentPos != pos):
            current = current.getNext()
            currentPos += 1

        if (current == None):
            # ran off end of list, position not in list
            return False

        # found position
        previous = current.getPrevious()
        if (previous == None):
            # popping first item from list
            self.head = current.getNext()
            if (self.head != None):
                self.head.setPrevious(None)
        else:
            following = current.getNext()
            previous.setNext(following)
            if (following != None):
                following.setPrevious(previous)
        return current.getData()
    


class SinglyLinkedList:

    def __init__(self):
        self.head = None

    def add(self, item):
        temp = Node(item)
        temp.setNext(self.head)
        self.head = temp

    def traverse(self):
        current = self.head
        while (current):
            print "->" + str(current.getData()),
            current = current.getNext()

    def remove(self, item):
        found = False
        current = self.head
        previous = None

        while (current != None and not found):
            if (current.getData() == item):
                found = False
            else:
                previous = current
                current = current.getNext()

        if (previous == None):
            # must delete first item
            self.head = current.getNext()
        elif (current == None):
            return False
        else:
            previous.setNext(current.getNext())

    def search(self, item):
        found = False
        current = self.head

        while (current != None and not found):
            if (current.getData() == item):
                found = True
            else:
                current = current.getNext()

        if (found):
            return True
        else:
            return False

    def isEmpty(self):
        return self.head == None

    def length(self):
        current = self.head
        count = 0

        while (current != None):
            count += 1
            current = current.getNext()

        return count

    def append(self, item):
        if (self.isEmpty()):
            self.add(item)
            return

        current = self.head
        previous = None
        
        while (current != None):
            previous = current
            current = current.getNext()

        temp = Node(item)
        previous.setNext(temp)

    def index(self, item):
        current = self.head
        pos = 0
        found = False

        while (current != None and not found):
            if (current.getData() == item):
                found = True
            else:
                pos += 1
                current = current.getNext()

        if (found):
            return pos
        else:
            return False

    def insert(self, pos, item):
        if (pos == 0):
            self.add(item)

        current = self.head
        previous = None
        currentPos = 0

        while (currentPos != pos):
            currentPos += 1
            previous = current
            current = current.getNext()
            
        temp = Node(item)
        previous.setNext(temp)
        temp.setNext(current)
        
    def pop(self, pos=None):
        if (pos == None):
            pos = self.length() - 1

        current = self.head
        previous = None
        currentPos = 0

        while (currentPos != pos):
            previous = current
            current = current.getNext()
            currentPos += 1

        if (previous == None):
            # popping first element from list
            self.head = current.getNext()
        else:
            previous.setNext(current.getNext())

        return current.getData()
        




def swap(LL, x, y):
    xNode = None
    yNode = None
    prevx = None
    prevy = None
    xafter = None
    yafter = None

    foundx = False
    foundy = False
    current = LL.head
    previous = None

    while (not foundx or not foundy):
        if (current.getData() == x):
            foundx = True
            xNode = current
            prevx = previous
            xafter = current.getNext()
        if (current.getData() == y):
            foundy = True
            yNode = current
            prevy = previous
            yafter = current.getNext()
        previous = current
        current = current.getNext()

    if (prevy == xNode):
        # nodes adjacent
        prevx.setNext(yNode)
        xNode.setNext(yafter)
        yNode.setNext(xNode)
    elif (prevx == yNode):
        # nodes adjacent
        prevy.setNext(xNode)
        yNode.setNext(xafter)
        xNode.setNext(yNode)
    elif (prevx == None):
        # x is head node
        xNode.setNext(yafter)
        yNode.setNext(xafter)
        prevy.setNext(xNode)
        LL.head = yNode
    elif (prevy == None):
        # y is head node
        xNode.setNext(yafter)
        yNode.setNext(xafter)
        prevx.setNext(yNode)
        LL.head = xNode
    else:
        prevx.setNext(yNode)
        xNode.setNext(yafter)
        yNode.setNext(xafter)
        prevy.setNext(xNode)

def printReverse(LL):
    if (LL == None):
        return
    printReverse(LL.getNext())
    print LL.getData()

def deleteDupsSorted(LL):
    current = LL.head.getNext()
    previous = LL.head
    while (current != None):
        if (previous.getData() == current.getData()):
            # delete current
            previous.setNext(current.getNext())
            current.setNext(None)
            current = previous.getNext()
        else:
            previous = current
            current = current.getNext()
    
    
        


l = SinglyLinkedList()
l.add(11)
#l.add(11)
#l.add(11)
#l.append(21)
#l.append(43)
#l.append(43)
#l.append(60)
l.traverse()
print ""
deleteDupsSorted(l)
l.traverse()

File: dataStructures/linked_list.rb
class Node

  def initialize(data)
    @data = data
    @next_node = nil
  end

  def get_data
    return @data
  end

  def get_next
    return @next_node
  end

  def set_data(data)
    @data = data
  end

  def set_next(node)
    @next_node = node
  end

end


class CircularLinkedList

  def initialize
    @head = nil
  end

  def add(item)
    if (@head.nil?)
      temp = Node.new(item)
      temp.set_next(temp)
      @head = temp
      return
    end

    start = @head
    current = @head.get_next
    previous = start

    while (current != start)
      previous = current
      current = current.get_next
    end

    temp = Node.new(item)
    temp.set_next(@head)
    @head = temp
    previous.set_next(temp)
  end

  def length
    if (is_empty?)
      return 0
    end

    count = 1 # at least one item
    start = @head
    current = @head.get_next

    while (current != start)
      count += 1
      current = current.get_next
    end

    return count
  end

  def is_empty?
    return @head.nil?
  end

  def remove(item)
    start = @head
    previous = start
    current = start.get_next
    found = false

    loop do
    #while (current != start and not found)
      if (current.get_data == item)
        found = true
      else
        previous = current
        current = current.get_next
      end
      break if (found or current == start)
    end

    if (not found)
      return false
    end


    if (current == start)
      # deleting first item
      if (previous == start)
        # deleting first item from length 1 list
        @head = nil
      else
        # deleting first item from length > 1 list
        @head = current.get_next
        previous.set_next(current.get_next)
      end
    else
      # remove like in singly linked list
      previous.set_next(current.get_next)
    end
  end

  def search(item)
    if (is_empty?)
      return false
    end

    start = @head
    current = start
    found = false
    
    loop do
      if (current.get_data == item)
        found = true
      else
        current = current.get_next
      end
      break if (found or current == start)
    end

    return found
  end

  def append(item)
    if (is_empty?)
      add(item)
      return
    end

    start = @head
    current = start
    previous = start
    
    loop do
      previous = current
      current = current.get_next
      break if (current == start)
    end

    temp = Node.new(item)
    temp.set_next(start)
    previous.set_next(temp)
  end

  def index(item)
    start = @head
    current = start
    pos = 0
    found = false

    loop do
      if (current.get_data == item)
        found = true
      else
        pos += 1
        current = current.get_next
      end
      break if (found or current == start)
    end
    
    if (found)
      return pos
    else
      return false
    end
  end

  def insert(pos, item)
    if (pos == 0)
      add(item)
      return
    end
    
    start = @head
    current = start
    previous = start
    current_pos = 0
    
    while (current_pos != pos)
      previous = current
      current = current.get_next
      current_pos += 1
    end

    temp = Node.new(item)
    temp.set_next(current)
    previous.set_next(temp)
  end

  def pop(pos=(length-1))
    start = @head
    current = start
    previous = start
    current_pos = 0

    while (current_pos != pos)
      previous = current
      current = current.get_next
      current_pos += 1
    end

    if (current == start)
      # popping first item
      if (@head.get_next == start)
        # popping first item from length 1 list
        @head = nil
        return start.get_data
      else
        # need to traverse to end to update end pointer
        current = start.get_next
        while (current != start)
          previous = current
          current = current.get_next
        end
        previous.set_next(start.get_next)
        @head = start.get_next
        return start.get_data
      end
    else
      previous.set_next(current.get_next)
      return current.get_data
    end
  end

end


begin
  l = CircularLinkedList.new
  l.add(12)
  l.add(13)
  l.add(15)
  puts l.inspect
  puts l.length
  puts l.pop(1)
  puts l.inspect
end
